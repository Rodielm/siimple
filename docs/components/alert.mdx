---
title: "Alert"
description: "Display block messages"
---

# Alert

An **alert** is a colored notification block generally used to catch the attention of your users of something important. An alert can be created adding the `siimple-alert` class to a `<div>` tag.

```html live=true
<div class="siimple-alert">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit...
</div>
```

You can add a close button to the alert using the `siimple-close` component:

```html live=true
<div class="siimple-alert">
    <div>Content of the alert message</div>
    <div class="siimple-close has-ml-auto"></div>
</div>
```

### Variants

Variants for this component can be defined in the `alert` section of the variants configuration:

```scss
@use "siimple" with (
    $variants: (
        "alert": (
            "default": (
                "background-color": "primary",
                "color": "white",
            ),
        ),
    ),
));
```
