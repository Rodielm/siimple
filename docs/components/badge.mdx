---
title: "Badge"
description: "Small labelling element"
---

# Badge

**Badges** are useful elements for indicating status, attach metadata to another element or displaying numbers for counts. You can easily turn a `<span>` or `<div>` element to a badge adding the `siimple-badge` class.

```html live=true
<span class="siimple-badge">Badge</span>
<span class="siimple-badge is-secondary">Badge</span>
```

### Variants

Variants for this component can be defined in the `badge` section of the theme configuration:

```scss
siimple.theme((
    "badge": (
        "primary": (
            "background-color": "primary",
            "color": "white",
        ),
        "secondary": (
            "background-color": "secondary",
            "color": "white",
        ),
    ),
));
```
