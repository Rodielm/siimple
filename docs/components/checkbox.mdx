---
title: "Checkbox"
description: "Checkbox input component"
---

# Checkbox

Style your checkbox elements adding the `siimple-checkbox` style to the `<input>` element. 

```html live=true
<label class="siimple-label" for="checkbox0">
    <span class="has-mr-2">Default checkbox</span>
    <input type="checkbox" id="checkbox0" class="siimple-checkbox" />
</label>
```

By default, the checkbox will be displayed unchecked, but you can add a **checked** attribute to the `<input>` element to display the checkbox as checked, or use the `is-active` modifier to manually display it as active.

```html live=true
<label class="siimple-label" for="checkbox1">
    <span class="has-mr-2">Active checkbox</span>
    <input type="checkbox" id="checkbox1" class="siimple-checkbox" checked />
</label>
```

Checkboxes can also display the **indeterminate** state. Remember that there is no HTML attribute to set this state, the only way to set this state is via [JavaScript](https://developer.mozilla.org/en-US/docs/Web/CSS/:indeterminate).

```html
<input type="checkbox" id="checkbox4" class="siimple-checkbox" />

<!-- Set the indeterminate state via javascript -->
<script type="text/javascript">
    document.getElementById("checkbox4").indeterminate = true;
</script>
```

### Variants

Variants for this component can be defined in the `form.checkbox` section of the variants configuration:

```scss
@use "siimple" with (
    $variants: (
        "form.checkbox": (
            "default": (
                "background-color": "fill",
                "color": "white",
                "&:checked, &:indeterminate": (
                    "background-color": "primary",
                ),
                "&:indeterminate:after": (
                    "border-bottom-color": "white",
                ),
                "&:checked:after": (
                    "border-bottom-color": "white",
                    "border-left-color": "white",
                ),
            ),
        ),
    ),
);
```
