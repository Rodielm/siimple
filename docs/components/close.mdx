---
title: "Close"
description: "A close button"
---

# Close

Use the `siimple-close` class to create a close button:

```html align=center live=true
<div class="siimple-close"></div>
```


### Variants

Variants for this component can be defined in the `button.close` section of the variants configuration:

```scss
@use "siimple" with (
    $variants: (
        "button.close": (
            "default": (
                "color": "currentColor",
            ),
        ),
    ),
);
```
