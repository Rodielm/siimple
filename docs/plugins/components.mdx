---
title: Components plugin
---

# Components Plugin

We provide a collection of built-in UI components that you can customize and reuse across your projects. We have divided our components collection into two groups: 

- **Core components**: primary and most common used components, that includes alerts, buttons, navigation links, and more!
- **Experimental components**: additional components that will be merged in a future with the core components.

**Components are always enabled by default**, so you do not need to import them as a plugin. In case that you do not need components in your output CSS, you can disable them using a **configuration flag**:

```scss
@use "siimple" with (
    $flags: (
        "add-core-components": false,  // Skip core components
        "add-experimental-components": false, // Skip experimental components 
    ),
);
```

Also, in case that you only need some components in your output CSS, you can import each component as a single plugin in the `$plugins` section of the configuration. Remember that you must import the `siimple/plugins` module for adding plugins to the configuration:

```scss
@use "siimple/plugins" as plugins;

@use "siimple" with (
    $flags: (
        "add-core-components": false,
        "add-experimental-components": false,
    ),
    $plugins: (
        plugins.btn(),
        plugins.column(),
        plugins.text(),
    ),
);
```


