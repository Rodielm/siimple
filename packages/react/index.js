import React from "react";
import {classNames, createCache, registerCss} from "@siimple/styled";

// Contexts
export const ThemeContext = React.createContext({});
export const CacheContext = React.createContext(
    createCache({key: "siimple-react"}),
);

// Internal css compiler
const __useCss = (css, isGlobal) => {
    const theme = React.useContext(ThemeContext);
    const cache = React.useContext(CacheContext);

    return registerCss(css, theme, cache, isGlobal, false);
};

// Available hooks
export const useTheme = () => React.useContext(ThemeContext);
export const useCss = css => __useCss(css, false);
export const useGlobalCss = css => __useCss(css, true);

// Theme provider component
export const ThemeProvider = props => {
    let theme = useTheme();
    if (props.theme && typeof props.theme === "object" && props.theme !== theme) {
        theme = props.theme;
    }
    return React.createElement(ThemeContext.Provider, {value: theme}, props.children);
};

// HOC that wraps a component and adds the current theme as a prop
export const withTheme = Component => {
    const name = Component.displayName || Component.name || "Component";
    const ThemedComponent = React.forwardRef((props, ref) => {
        const theme = useTheme();

        return React.createElement(Component, {
            ...props,
            theme: theme,
            ref: ref,
        });
    });
    ThemedComponent.displayName = `withTheme(${name})`;
    return ThemedComponent;
};

// Create an styled element
export const styled = (type = "div", css) => {
    return React.forwardRef((props, ref) => {
        const {css: cssProp, ...rest} = props;
        const className = useCss({
            ...css,
            ...(cssProp || {}),
        });

        return React.createElement(type, {
            ...rest,
            ref: ref,
            className: classNames(props.className, className),
        });
    });
};
