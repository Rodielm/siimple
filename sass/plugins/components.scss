// @description Load essential components of siimple
@mixin all-components () {
    @include alert();
    @include badge();
    @include btn();
    @include card();
    @include checkbox();
    @include close();
    @include column();
    @include content();
    @include crumb();
    @include divider();
    @include input();
    @include label();
    @include navlink();
    @include paragraph();
    @include quote();
    @include radio();
    @include select();
    @include spinner();
    @include switch();
    @include tab();
    @include table();
    @include text();
    @include textarea();
    @include title();
}
