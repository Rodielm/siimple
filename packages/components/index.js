import React from "react";
import {useTheme, Box, styled} from "@siimple/react";
import {elements} from "@siimple/modules/elements.js";
import {markup} from "@siimple/modules/markup.js";

// Base styles
export const BaseStyles = props => {
    const theme = useTheme();
    return React.createElement(Box, {
        ...props,
        as: "div",
        css: {
            boxSizing: "border-box",
            background: "background",
            color: "text",
            fontFamily: "body",
            fontWeight: "body",
            lineHeight: "body",
            ...(theme.root || {}),
            ...(props.css || {}),
        },
    });
};

// Generate elements
export const Elements = Object.fromEntries(
    Object.keys(elements).map(key => {
        const config = elements[key];
        const Component = React.forwardRef((props, ref) => {
            const {variant, ...rest} = props;
            const theme = useTheme();
            const variantCss = {};
    
            // Assign variants
            if (config?.variants && variant) {
                Object.assign(variantCss, {
                    ...(config?.defaultVariants?.[variant] || {}),
                    ...(theme[config.variants]?.[variant] || {}),
                });
            }

            return React.createElement(Box, {
                ...(config?.defaultProps || {}),
                ...rest,
                as: props.as || config.type || "div",
                ref: ref,
                css: {
                    ...config?.styles,
                    ...variantCss,
                    ...props.css,
                },
            });
        });
        return [key, Component];
    }),
);

// Markup styled elements
export const Markup = Object.fromEntries(
    Object.keys(markup).map(key => {
        const Component = styled(key, {
            ...markup[key],
            apply: `styles.${key}`,
        });

        return [key, Component];   
    }),
);
