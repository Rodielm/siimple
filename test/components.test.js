/**
 * @jest-environment jsdom
 */

import renderer from "react-test-renderer";
import {ThemeProvider} from "@siimple/react";
import {Box, Alert} from "@siimple/components";

const selector = `style[data-siimple="siimple-react"]`;

describe("Box", () => {
    it("should render", () => {
        const component = renderer.create((
            <ThemeProvider theme={{}}>
                <Box css={{color: "white"}}>Hello world</Box>
            </ThemeProvider>
        ));

        expect(component.toJSON()).toMatchSnapshot();
    });
});

describe("All components", () => {
    it("should render", () => {
        const component = renderer.create((
            <ThemeProvider theme={{}}>
                <Alert>Hello world</Alert>
            </ThemeProvider>
        ));

        expect(component.toJSON()).toMatchSnapshot();
    });
});
