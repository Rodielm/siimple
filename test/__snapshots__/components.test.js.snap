// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All components should render 1`] = `
<div
  className="sii-358420926"
>
  Hello world
</div>
`;

exports[`Box should render 1`] = `
<div
  className="sii-1628842713"
>
  Hello world
</div>
`;
